src/relocate.rs

```diff
@@ -0,0 +1,81 @@
+// Copyright 2025 Manos Pitsidianakis <manos.pitsidianakis@linaro.org>
+// This project is dual-licensed under Apache 2.0 and MIT terms.
+// See LICENSE-APACHE and LICENSE-MIT for details.
+
+type Elf64Addr = u64;
+type Elf64Xword = u64;
+type Elf64Sxword = i64;
+
+const R_AARCH64_RELATIVE: u32 = 1027;
+
+macro_rules! elf64_r_type {
+    ($i:expr) => {{ (($i) & 0xffffffff) as u32 }};
+}
+
+#[derive(Copy, Clone)]
+#[repr(C)]
+struct Elf64Rela {
+    r_offset: Elf64Addr,
+    r_info: Elf64Xword,
+    r_addend: Elf64Sxword,
+}
+
+unsafe extern "C" {
+    static __rela_start: *mut Elf64Rela;
+    static __rela_end: *mut Elf64Rela;
+}
+
+/// # Safety
+///
+/// This function is unsafe because it modifies executable memory.
+///
+/// In order for this function to work correctly, we make the following assumptions:
+///
+/// - `offset` argument is the actual address the binary was loaded into memory
+/// - The binary uses rela relocations only (not relr)
+/// - The binary must be PIE (`-pie` linker option)
+/// - The binary's origin address is 0x0
+///
+/// Furthermore, you must also generate relocations for read-only text segments (linker option
+/// `-Z notext`).
+pub unsafe extern "C" fn relocate_image(offset: u64) {
```

Comment:
How do we know that this function is safe to run before applying relocations?

Reply:
Replaced with assembly implementation